pymtl3/dsl/Component.py

```diff
@@ -714,19 +723,113 @@ def replace_component_with_obj( top, foo, new_obj, check=True ):
     if check:
       top.check()
 
-  def add_connection( s, o1, o2 ):
-    try:
-      top = s._dsl.elaborate_top
-    except AttributeError:
-      raise NotElaboratedError()
+  def add_value_port( top, parent, name, o ):
+    top._check_called_at_elaborate_top( "add_port" )
 
-    try:
-      s._connect( o1, o2, internal=False )
-    except AssertionError as e:
-      raise InvalidConnectionError( "\n{}".format(e) )
+    assert isinstance( o, (InPort, OutPort) )
+    # If we are adding field parent.x, we simply reuse the setattr hook
+    assert not hasattr( parent, name ), f"Invalid add_value_port call: {parent} already has field {name}!"
 
-    for x, adjs in s._dsl.adjacency.items():
-      top._dsl.all_adjacency[x].update( adjs )
+    NamedObject._elaborate_stack = [ parent ]
+    NamedObject.__setattr__ = NamedObject.__setattr_for_elaborate__
+    setattr( parent, name, o )
+    del NamedObject.__setattr__
+    del NamedObject._elaborate_stack
+
+    top._dsl.all_signals.add( o )
+    top._dsl.all_named_objects.add( o )
+
+  def add_connection( top, o1, o2 ):
+
+    # Currently only support connecting signals
+
+    top._check_called_at_elaborate_top( "add_connection" )
+    if isinstance( o2, Connectable ): o1, o2 = o2, o1
+
+    assert isinstance( o1, Connectable ), "Cannot connect two non-connectable objects"
+
+    assert isinstance( o1, Signal ), "Currently only support connecting signals/constants"
```

Comment:
I wonder if it is possible (eventually) to connect a lambda block to a signal.. it's basically adding another concurrent process to the component and I'm sure designers will appreciate this post-elaboration flexibility.

Reply:
This is a very good point. The question is there is no "s" to catch. Basically we need the user to provide raw string as source code.